### ISSUE TITLE
Lb1

### ISSUE THREAD

**Issue by Homka3310** [Homka3310/LearningMachine#1]
import os
import re
import string
import numpy as np
import pickle
from collections import Counter
import nltk
from nltk.corpus import twitter_samples, stopwords
from nltk.tokenize import TweetTokenizer
from nltk.stem import PorterStemmer
from sklearn.metrics import accuracy_score, precision_recall_fscore_support, confusion_matrix
import pandas as pd

def load_twitter_samples():
    all_pos = twitter_samples.strings('positive_tweets.json')
    all_neg = twitter_samples.strings('negative_tweets.json')
    train_pos = all_pos[:4000]
    test_pos = all_pos[4000:]
    train_neg = all_neg[:4000]
    test_neg = all_neg[4000:]
    train_x = train_pos + train_neg
    test_x = test_pos + test_neg
    train_y = np.append(np.ones(len(train_pos)), np.zeros(len(train_neg)))
    test_y = np.append(np.ones(len(test_pos)), np.zeros(len(test_neg)))
    return train_x, train_y.reshape((-1,1)), test_x, test_y.reshape((-1,1))

def load_csv(path, text_col='text', label_col='label', sep='\t'):
    df = pd.read_csv(path, sep=sep)
    df = df[[text_col, label_col]].dropna()
    texts = df[text_col].astype(str).tolist()
    labels = df[label_col].astype(int).tolist()
    split = int(len(texts) * 0.8)
    train_x = texts[:split]
    test_x = texts[split:]
    train_y = np.array(labels[:split]).reshape((-1,1))
    test_y = np.array(labels[split:]).reshape((-1,1))
    return train_x, train_y, test_x, test_y

tokenizer = TweetTokenizer(preserve_case=False, strip_handles=True, reduce_len=True)
stemmer = PorterStemmer()
stopwords_english = stopwords.words('english')

def process_tweet(tweet):
    tweet = re.sub(r'\$\w*', '', tweet)
    tweet = re.sub(r'^RT[\s]+', '', tweet)
    tweet = re.sub(r'https?://[^\s\n\r]+', '', tweet)
    tweet = re.sub(r'#', '', tweet)
    tokens = tokenizer.tokenize(tweet)
    tweets_clean = []
    for word in tokens:
        if (word not in stopwords_english and word not in string.punctuation):
            stem_word = stemmer.stem(word)
            tweets_clean.append(stem_word)
    return tweets_clean

def build_freqs(tweets, ys):
    yslist = np.squeeze(ys).tolist()
    freqs = {}
    for y, tweet in zip(yslist, tweets):
        for word in process_tweet(tweet):
            pair = (word, float(y))
            freqs[pair] = freqs.get(pair, 0) + 1
    return freqs

def sigmoid(z):
    return 1.0 / (1.0 + np.exp(-z))

def gradientDescent(X, y, theta, alpha, num_iters, reg_lambda=0.0, verbose=False):
    m = X.shape[0]
    for i in range(num_iters):
        z = np.dot(X, theta)
        h = sigmoid(z)
        J = (-1/m) * (np.dot(y.T, np.log(h + 1e-12)) + np.dot((1-y).T, np.log(1-h + 1e-12)))
        grad = (1/m) * np.dot(X.T, (h - y))
        reg = (reg_lambda / m) * theta
        reg[0] = 0
        theta = theta - alpha * (grad + reg)
        if verbose and (i % max(1, num_iters//10) == 0):
            print(f"Iter {i}/{num_iters} - J = {float(J):.6f}")
    return float(J), theta

def extract_features(tweet, freqs):
    word_l = process_tweet(tweet)
    x = np.zeros((1, 3))
    x[0, 0] = 1
    for word in word_l:
        x[0, 1] += freqs.get((word, 1.0), 0)
        x[0, 2] += freqs.get((word, 0.0), 0)
    return x

def predict_tweet(tweet, freqs, theta):
    x = extract_features(tweet, freqs)
    y_pred = sigmoid(np.dot(x, theta))
    return y_pred

def test_logistic_regression(test_x, test_y, freqs, theta):
    y_hat = []
    for tweet in test_x:
        y_pred = predict_tweet(tweet, freqs, theta)
        if y_pred > 0.5:
            y_hat.append(1)
        else:
            y_hat.append(0)
    accuracy = (np.sum(np.array(y_hat) == np.squeeze(test_y))) / len(test_x)
    return accuracy, np.array(y_hat)

def main(use_twitter=True, csv_path=None, save_dir='output', alpha=1e-9, num_iters=1500, reg_lambda=0.0, verbose=True):
    os.makedirs(save_dir, exist_ok=True)
    if use_twitter:
        train_x, train_y, test_x, test_y = load_twitter_samples()
    else:
        if csv_path is None:
            raise ValueError('Якщо use_twitter=False, передайте шлях до csv через csv_path')
        train_x, train_y, test_x, test_y = load_csv(csv_path)
    if verbose:
        print(f"Train size: {len(train_x)}, Test size: {len(test_x)}")
    freqs = build_freqs(train_x, train_y)
    with open(os.path.join(save_dir, 'freqs.pkl'), 'wb') as f:
        pickle.dump(freqs, f)
    if verbose:
        print('freqs збережено в', os.path.join(save_dir, 'freqs.pkl'))
    X = np.zeros((len(train_x), 3))
    for i in range(len(train_x)):
        X[i, :] = extract_features(train_x[i], freqs)
    Y = train_y
    theta = np.zeros((3, 1))
    J, theta = gradientDescent(X, Y, theta, alpha, num_iters, reg_lambda=reg_lambda, verbose=verbose)
    if verbose:
        print(f"Помилка після навчання: {J:.8f}")
        print("Вектор вагових коефіцієнтів:", [round(t,8) for t in np.squeeze(theta)])
    acc, y_hat = test_logistic_regression(test_x, test_y, freqs, theta)
    prec, rec, f1, _ = precision_recall_fscore_support(np.squeeze(test_y), y_hat, average='binary')
    if verbose:
        print(f"Точність на тесті: {acc:.4f}")
        print(f"Precision: {prec:.4f}, Recall: {rec:.4f}, F1: {f1:.4f}")
        print('Confusion matrix:\n', confusion_matrix(np.squeeze(test_y), y_hat))
    with open(os.path.join(save_dir, 'theta.pkl'), 'wb') as f:
        pickle.dump(theta, f)
    report_md = os.path.join(save_dir, 'report_summary.md')
    with open(report_md, 'w', encoding='utf-8') as f:
        f.write('# Lab 1 — Sentiment analysis using logistic regression\n')
        f.write(f'## Датасет: "twitter_samples" (NLTK), train={len(train_x)}, test={len(test_x)}\n')
        f.write('## Гіперпараметри:\n')
        f.write(f'- alpha = {alpha}\n')
        f.write(f'- num_iters = {num_iters}\n')
        f.write(f'- reg_lambda = {reg_lambda}\n')
        f.write('\n')
        f.write('## Результати:\n')
        f.write(f'- Loss after training: {J:.8f}\n')
        f.write(f'- Test accuracy: {acc:.4f}\n')
        f.write(f'- Precision: {prec:.4f}\n')
        f.write(f'- Recall: {rec:.4f}\n')
        f.write(f'- F1: {f1:.4f}\n')
    if verbose:
        print('Звіт збережено в', report_md)
    readme = os.path.join(save_dir, 'README_for_GitHub.md')
    with open(readme, 'w', encoding='utf-8') as f:
        f.write('# Lab1 - Logistic regression for sentiment analysis\n')
        f.write('Repository contains:\n')
        f.write('- lab1_sentiment_logreg.py (this script)\n')
        f.write('- output/freqs.pkl, theta.pkl, report_summary.md\n')
        f.write('\n## How to run\n')
        f.write('1. pip install -r requirements.txt\n')
        f.write('2. python lab1_sentiment_logreg.py\n')
        f.write('\nNotes:\n- If you use twitter_samples you must download NLTK data:\n')
        f.write("  import nltk\n  nltk.download('twitter_samples')\n  nltk.download('stopwords')\n")
    req = os.path.join(save_dir, 'requirements.txt')
    with open(req, 'w', encoding='utf-8') as f:
        f.write('numpy\n')
        f.write('nltk\n')
        f.write('scikit-learn\n')
        f.write('pandas\n')
    if verbose:
        print('README і requirements збережено в', save_dir)
    examples = [
        'I love this movie. It was amazing and fantastic!',
        'This is the worst movie I have ever seen. Totally boring.',
        'Not bad, but I expected more.'
    ]
    print('\nПриклади передбачень:')
    for ex in examples:
        prob = predict_tweet(ex, freqs, theta)
        print(ex)
        print(f'P(positive) = {float(prob):.4f} ->', 'positive' if prob>0.5 else 'negative')
    return {
        'freqs_path': os.path.join(save_dir, 'freqs.pkl'),
        'theta_path': os.path.join(save_dir, 'theta.pkl'),
        'report': report_md,
        'readme': readme,
        'requirements': req
    }

if __name__ == '__main__':
    artifacts = main(use_twitter=True, csv_path=None, save_dir='output', alpha=1e-9, num_iters=1500, reg_lambda=0.0, verbose=True)
    print('\nГотово. Артефакти:', artifacts)
